In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_excel('бд.xlsx')
df.head()

Unnamed: 0            Исходные данные Исходные данные.1 Исходные данные.2  \
0      № п/п          Координата яндекс  Название посёлка         Поселение   
1          1  55.614630070,37.232309183         кв-л № 73        Внуковское   
2          2  55.614013325,37.230067327         кв-л № 73        Внуковское   
3          3  55.614692112,37.230466380         кв-л № 73        Внуковское   
4          4  55.615425187,37.229852989         кв-л № 73        Внуковское   

                  Добавила               Добавила.1  \
0  Поселение\n(без Москвы)  Плодадь поселения (км²)   
1               Внуковское                     25.6   
2               Внуковское                     25.6   
3               Внуковское                     25.6   
4               Внуковское                     25.6   

                                          Добавила.2  \
0  Удаленность от Москвы (км). Точка отсчета цент...   
1                                               32.4   
2                                               32.4   
3                                               32.4   
4                                               32.4   

                                          Добавила.3  \
0  Статус населенного пункта (СНТ/ОНТ/ДНТ или гор...   
1                                                СНТ   
2                                                СНТ   
3                                                СНТ   
4                                                СНТ   

                        Добавила.4 Исходные данные.3 Исходные данные.4  \
0  Постоянное/временное проживание     Кад. номер ЗУ    Кад. номер ОКС   
1             временное проживание                 -                 -   
2             временное проживание                 -                 -   
3             временное проживание                 -                 -   
4             временное проживание                 -                 -   

                  Исходные данные.5  
0  Снят/необходимо снять с контроля  
1                               NaN  
2                               NaN  
3                               NaN  
4                               NaN

In [4]:
df = pd.read_excel('испрБД.xlsx')
df.head()

№ п/п          Координата яндекс Название посёлка   Поселение  \
0     1  55.614630070,37.232309183        кв-л № 73  Внуковское   
1     2  55.614013325,37.230067327        кв-л № 73  Внуковское   
2     3  55.614692112,37.230466380        кв-л № 73  Внуковское   
3     4  55.615425187,37.229852989        кв-л № 73  Внуковское   
4     5  55.617217454,37.268645722   деревня Ликова  Внуковское   

  Поселение\n(без Москвы) Плодадь поселения (км²)  \
0              Внуковское                    25.6   
1              Внуковское                    25.6   
2              Внуковское                    25.6   
3              Внуковское                    25.6   
4              Внуковское                       -   

  Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)  \
0                                               32.4                         
1                                               32.4                         
2                                               32.4                         
3                                               32.4                         
4                                                  -                         

  Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)  \
0                                                СНТ          
1                                                СНТ          
2                                                СНТ          
3                                                СНТ          
4                           проблематика отсутствует          

  Постоянное/временное проживание Кад. номер ЗУ Кад. номер ОКС  \
0            временное проживание             -              -   
1            временное проживание             -              -   
2            временное проживание             -              -   
3            временное проживание             -              -   
4                               -             +              +   

  Снят/необходимо снять с контроля  
0                              NaN  
1                              NaN  
2                              NaN  
3                              NaN  
4                 Снято с контроля

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24698 entries, 0 to 24697
Data columns (total 12 columns):
 #   Column                                                                    Non-Null Count  Dtype 
---  ------                                                                    --------------  ----- 
 0   № п/п                                                                     24698 non-null  object
 1   Координата яндекс                                                         24698 non-null  object
 2   Название посёлка                                                          23900 non-null  object
 3   Поселение                                                                 24698 non-null  object
 4   Поселение
(без Москвы)                                                    24698 non-null  object
 5   Плодадь поселения (км²)                                                   24698 non-null  object
 6   Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший пут

Видно что у нас только в 2 столбцах нет ифнормации: 'Название поселка', 'Снят/необходимо снять с контроля'

In [6]:
# Посмотрим на строки, где 'Название посёлка' имеет значение NaN
df[df['Название посёлка'].isnull()]

№ п/п Координата яндекс Название посёлка  \
23823                    23835                 -              NaN   
23824                    23836                 -              NaN   
23825                    23837                 -              NaN   
23826                    23838                 -              NaN   
23827                    23839                 -              NaN   
...                        ...               ...              ...   
24693  Внеплан. Список от ДЭПР                 -              NaN   
24694  Внеплан. Список от ДЭПР                 -              NaN   
24695  Внеплан. Список от ДЭПР                 -              NaN   
24696  Внеплан. Список от ДЭПР                 -              NaN   
24697  Внеплан. Список от ДЭПР                 -              NaN   

                   Поселение Поселение\n(без Москвы) Плодадь поселения (км²)  \
23823  САО Западное Дегунино                  Москва                       -   
23824  САО Западное Дегунино                  Москва                       -   
23825  САО Западное Дегунино                  Москва                       -   
23826  САО Западное Дегунино                  Москва                       -   
23827            ЮЗАО Зюзино                  Москва                       -   
...                      ...                     ...                     ...   
24693             Кленовское              Кленовское                  117.72   
24694        Краснопахорское         Краснопахорское                   87.78   
24695          Воскресенское           Воскресенское                   24.82   
24696            Рязановское             Рязановское                   40.48   
24697        Краснопахорское         Краснопахорское                   87.78   

      Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)  \
23823                                                  -                         
23824                                                  -                         
23825                                                  -                         
23826                                                  -                         
23827                                                  -                         
...                                                  ...                         
24693                                               67.2                         
24694                                               48.1                         
24695                                               34.7                         
24696                                               35.4                         
24697                                               48.1                         

      Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)  \
23823                                                  -          
23824                                                  -          
23825                                                  -          
23826                                                  -          
23827                                                  -          
...                                                  ...          
24693                                                снт          
24694                                            деревня          
24695                                            деревня          
24696                                                снт          
24697                                            деревня          

      Постоянное/временное проживание Кад. номер ЗУ Кад. номер ОКС  \
23823                               -             -              -   
23824                               -             +              +   
23825                               -             -              -   
23826                               -             +              +   
23827                               -             +              +   
...                      

In [7]:
# заменю неизвестные нам послеки на 'No name X', от 1 до 798: (Не смогу их дропнуть тк тогда анализ сильно изменится - пропадут 798 поселков)

new_names = ['Безымянный ' + str(i) for i in range(1, 799)]
nan_settlements_index = df[df['Название посёлка'].isnull()].index
df.loc[nan_settlements_index, 'Название посёлка'] = new_names

df.loc[nan_settlements_index].head()


№ п/п Координата яндекс Название посёлка              Поселение  \
23823  23835                 -     Безымянный 1  САО Западное Дегунино   
23824  23836                 -     Безымянный 2  САО Западное Дегунино   
23825  23837                 -     Безымянный 3  САО Западное Дегунино   
23826  23838                 -     Безымянный 4  САО Западное Дегунино   
23827  23839                 -     Безымянный 5            ЮЗАО Зюзино   

      Поселение\n(без Москвы) Плодадь поселения (км²)  \
23823                  Москва                       -   
23824                  Москва                       -   
23825                  Москва                       -   
23826                  Москва                       -   
23827                  Москва                       -   

      Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)  \
23823                                                  -                         
23824                                                  -                         
23825                                                  -                         
23826                                                  -                         
23827                                                  -                         

      Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)  \
23823                                                  -          
23824                                                  -          
23825                                                  -          
23826                                                  -          
23827                                                  -          

      Постоянное/временное проживание Кад. номер ЗУ Кад. номер ОКС  \
23823                               -             -              -   
23824                               -             +              +   
23825                               -             -              -   
23826                               -             +              +   
23827                               -             +              +   

      Снят/необходимо снять с контроля  
23823                              NaN  
23824                              NaN  
23825                              NaN  
23826                              NaN  
23827                              NaN

Теперь займусь пропусками с колонки 'Снят/необходимо снять с контроля':

In [8]:
df['Снят/необходимо снять с контроля'].unique()

array([nan, 'Снято с контроля', 'Некапитальные. Направлено на снятие'],
      dtype=object)

In [9]:
# Заменим пропуски на 'Нет информации':
new_names = ['Нет информации']
nan_settlements_index = df[df['Снят/необходимо снять с контроля'].isnull()].index
df.loc[nan_settlements_index, 'Снят/необходимо снять с контроля'] = new_names

df.loc[nan_settlements_index].head()

№ п/п          Координата яндекс Название посёлка   Поселение  \
0     1  55.614630070,37.232309183        кв-л № 73  Внуковское   
1     2  55.614013325,37.230067327        кв-л № 73  Внуковское   
2     3  55.614692112,37.230466380        кв-л № 73  Внуковское   
3     4  55.615425187,37.229852989        кв-л № 73  Внуковское   
5     6  55.617449028,37.269484212   деревня Ликова  Внуковское   

  Поселение\n(без Москвы) Плодадь поселения (км²)  \
0              Внуковское                    25.6   
1              Внуковское                    25.6   
2              Внуковское                    25.6   
3              Внуковское                    25.6   
5              Внуковское                    25.6   

  Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)  \
0                                               32.4                         
1                                               32.4                         
2                                               32.4                         
3                                               32.4                         
5                                               32.4                         

  Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)  \
0                                                СНТ          
1                                                СНТ          
2                                                СНТ          
3                                                СНТ          
5                                            деревня          

  Постоянное/временное проживание Кад. номер ЗУ Кад. номер ОКС  \
0            временное проживание             -              -   
1            временное проживание             -              -   
2            временное проживание             -              -   
3            временное проживание             -              -   
5           постоянное проживание             -              +   

  Снят/необходимо снять с контроля  
0                   Нет информации  
1                   Нет информации  
2                   Нет информации  
3                   Нет информации  
5                   Нет информации

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24698 entries, 0 to 24697
Data columns (total 12 columns):
 #   Column                                                                    Non-Null Count  Dtype 
---  ------                                                                    --------------  ----- 
 0   № п/п                                                                     24698 non-null  object
 1   Координата яндекс                                                         24698 non-null  object
 2   Название посёлка                                                          24698 non-null  object
 3   Поселение                                                                 24698 non-null  object
 4   Поселение
(без Москвы)                                                    24698 non-null  object
 5   Плодадь поселения (км²)                                                   24698 non-null  object
 6   Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший пут

### С пропусками разобрались, теперь посмотрим на значения в столбцах (мб там тоже есть какие-либо неточности/ошибки)

In [11]:
list(df.columns)

['№ п/п',
 'Координата яндекс',
 'Название посёлка',
 'Поселение',
 'Поселение\n(без Москвы)',
 'Плодадь поселения (км²)',
 'Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)',
 'Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)',
 'Постоянное/временное проживание',
 'Кад. номер ЗУ',
 'Кад. номер ОКС',
 'Снят/необходимо снять с контроля']

In [12]:
df['Поселение'].unique()

array(['Внуковское', 'Вороновское', 'Воскресенское', 'город Щербинка',
       'Десеновское', 'Киевский', 'Кленовское', 'Кокошкино',
       'Краснопахорское', 'Марушкинское', 'Михайлово-Ярцевское',
       'Московский', 'Мосрентген', 'Новофедоровское', 'Первомайское',
       'Роговское', 'Рязановское', 'Сосенское', 'город Троицк',
       'Филимонковское', 'Щаповское', 'САО Западное Дегунино',
       'ЮЗАО Зюзино', 'ЗАО Ново-Переделкино', 'ВАО Сокольники',
       'ЮВАО Некрасовка', 'ВАО Новокосино', 'ЮВАО Выхино-Жулебино',
       'СЗАО Щукино', 'ЮАО Царицыно', 'ВАО Богородское',
       'ЮАО Орехово-Борисово', 'ЗАО Внуково', 'СВАО Лианозово',
       'ВАО Новогиреево', 'САО Ховрино', 'СВАО Северный',
       'САО Молжаниновский', 'СЗАО Хорошево-Мневники',
       'ЮЗАО Южное Бутово', 'САО Хорошевский', 'ЗАО Проспект Вернадского',
       'ЮЗАО Ломоносовский', 'САО Аэропорт', 'ЗАО Солнцево',
       'СЗАО Куркино', 'ЮЗАО Северное Бутово', 'САО Войковский',
       'ВАО Косино-Ухтомский', 'ЮЗАО Га

In [13]:
df['Поселение\n(без Москвы)'].unique()

array(['Внуковское', 'Вороновское', 'Воскресенское', 'город Щербинка',
       'Десеновское', 'Киевский', 'Кленовское', 'Кокошкино',
       'Краснопахорское', 'Марушкинское', 'Михайлово-Ярцевское',
       'Московский', 'Мосрентген', 'Новофедоровское', 'Первомайское',
       'Роговское', 'Рязановское', 'Сосенское', 'город Троицк',
       'Филимонковское', 'Щаповское', 'Москва'], dtype=object)

In [14]:
df['Плодадь поселения (км²)'].unique()

array([25.6, ' -', 206.25, 24.82, 7.53, 52.99, 60.09, 117.72, 8.28, 87.78,
       50.63, 64.61, 40.39, 6.48, 150.8, 118.92, 178.23, 40.48, 66.81,
       139.2, 35.77, 86.88, '-'], dtype=object)

In [15]:
df['Снят/необходимо снять с контроля'].unique()

array(['Нет информации', 'Снято с контроля',
       'Некапитальные. Направлено на снятие'], dtype=object)

In [16]:
df['Постоянное/временное проживание'].unique()

array(['временное проживание', ' -', 'постоянное проживание',
       'Постоянное проживание', '-'], dtype=object)

In [17]:
# заменим теперь эти некорректности данных
df['Постоянное/временное проживание'] = df['Постоянное/временное проживание'].replace('Постоянное проживание', 'постоянное проживание')
df['Постоянное/временное проживание'] = df['Постоянное/временное проживание'].replace(' -', '-')

unique_values_after_replacement = df['Постоянное/временное проживание'].unique()
unique_values_after_replacement


array(['временное проживание', '-', 'постоянное проживание'], dtype=object)

In [18]:
df['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'].unique()

array(['СНТ', 'проблематика отсутствует', 'деревня', 'посёлок', 'поселок',
       'ТИЗ', 'ДСК', 'ТСН', 'ДНП', 'ЖСиЭК', 'ЖСПК', 'ЖСПКИЗ', 'село',
       'СПК', 'КИЗ', 'ТСЖ', 'город', 'рабочий посёлок', 'ДНТ', 'ДПК',
       'ДСПК', 'дачный посёлок, дачи', 'ДП', 'СТ', 'хутор', 'ЖСКИЗ',
       'ОНТ', 'СДТ', 'СК', 'ОПХ', 'ЖСК', 'CНТ', 'ИЖСК', 'ТОО', 'НСТ',
       'ПК', 'поселение Сосенское', '-', 'дередня', 'с/т', 'снт', 'ст'],
      dtype=object)

In [19]:
# заменим теперь эти некорректности данных
df['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'] = df['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'].replace('снт', 'СНТ')
df['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'] = df['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'].replace('ст', 'СТ')
df['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'] = df['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'].replace('СНТ', 'CНТ')
df['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'] = df['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'].replace('поселок', 'посёлок')
df['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'] = df['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'].replace('дередня', 'деревня')

unique_values_after_replacement = df['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'].unique()
unique_values_after_replacement


array(['CНТ', 'проблематика отсутствует', 'деревня', 'посёлок', 'ТИЗ',
       'ДСК', 'ТСН', 'ДНП', 'ЖСиЭК', 'ЖСПК', 'ЖСПКИЗ', 'село', 'СПК',
       'КИЗ', 'ТСЖ', 'город', 'рабочий посёлок', 'ДНТ', 'ДПК', 'ДСПК',
       'дачный посёлок, дачи', 'ДП', 'СТ', 'хутор', 'ЖСКИЗ', 'ОНТ', 'СДТ',
       'СК', 'ОПХ', 'ЖСК', 'ИЖСК', 'ТОО', 'НСТ', 'ПК',
       'поселение Сосенское', '-', 'с/т'], dtype=object)

In [20]:
df['Постоянное/временное проживание'].unique()

array(['временное проживание', '-', 'постоянное проживание'], dtype=object)

In [21]:
df['Плодадь поселения (км²)'].unique()

array([25.6, ' -', 206.25, 24.82, 7.53, 52.99, 60.09, 117.72, 8.28, 87.78,
       50.63, 64.61, 40.39, 6.48, 150.8, 118.92, 178.23, 40.48, 66.81,
       139.2, 35.77, 86.88, '-'], dtype=object)

In [22]:
df['Плодадь поселения (км²)'] = df['Плодадь поселения (км²)'].replace(' -', '-')
df['Плодадь поселения (км²)'].unique()

array([25.6, '-', 206.25, 24.82, 7.53, 52.99, 60.09, 117.72, 8.28, 87.78,
       50.63, 64.61, 40.39, 6.48, 150.8, 118.92, 178.23, 40.48, 66.81,
       139.2, 35.77, 86.88], dtype=object)

In [23]:
df['Кад. номер ЗУ'].unique()

array(['-', ' +'], dtype=object)

In [24]:
df['Кад. номер ЗУ'] = df['Кад. номер ЗУ'].replace(' +', '+')
df['Кад. номер ЗУ'].unique()

array(['-', '+'], dtype=object)

In [25]:
df['Кад. номер ОКС'].unique()

array(['-', ' +', '-\n', '- '], dtype=object)

In [26]:
df['Кад. номер ОКС'] = df['Кад. номер ОКС'].replace('- ', '-')
df['Кад. номер ОКС'] = df['Кад. номер ОКС'].replace(' +', '+')
df['Кад. номер ОКС'] = df['Кад. номер ОКС'].replace('-\n', '-')
df['Кад. номер ОКС'].unique()

array(['-', '+'], dtype=object)

In [27]:
df['Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)'].unique()

array([32.4, ' -', 62.9, 34.7, 32.5, 37.8, 64.4, 67.2, 41, 48.1, 40, 56.9,
       28.4, 24.1, 58.2, 51.4, 78.5, 35.4, 26.6, 41.9, 35.1, 58.9, '-'],
      dtype=object)

In [28]:
df['Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)'] = df['Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)'].replace(' -', '-')
df['Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)'].unique()

array([32.4, '-', 62.9, 34.7, 32.5, 37.8, 64.4, 67.2, 41, 48.1, 40, 56.9,
       28.4, 24.1, 58.2, 51.4, 78.5, 35.4, 26.6, 41.9, 35.1, 58.9],
      dtype=object)

### Я полностью обработала наш dataframe, теперь начнем проверять гипотезы:

# Гипотеза №1
Есть взаимосвязь между расстоянием от старой Москвы и тем оформлен ли объект недвижимости. Чем больше расстояние, тем выше шансы, что объект не оформлен. 

+ Зависимая переменная: оформлен/не оформлен
+ Предиктор: расстояние от старой Москвы


In [29]:
df.head()

№ п/п          Координата яндекс Название посёлка   Поселение  \
0     1  55.614630070,37.232309183        кв-л № 73  Внуковское   
1     2  55.614013325,37.230067327        кв-л № 73  Внуковское   
2     3  55.614692112,37.230466380        кв-л № 73  Внуковское   
3     4  55.615425187,37.229852989        кв-л № 73  Внуковское   
4     5  55.617217454,37.268645722   деревня Ликова  Внуковское   

  Поселение\n(без Москвы) Плодадь поселения (км²)  \
0              Внуковское                    25.6   
1              Внуковское                    25.6   
2              Внуковское                    25.6   
3              Внуковское                    25.6   
4              Внуковское                       -   

  Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)  \
0                                               32.4                         
1                                               32.4                         
2                                               32.4                         
3                                               32.4                         
4                                                  -                         

  Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)  \
0                                                CНТ          
1                                                CНТ          
2                                                CНТ          
3                                                CНТ          
4                           проблематика отсутствует          

  Постоянное/временное проживание Кад. номер ЗУ Кад. номер ОКС  \
0            временное проживание             -              -   
1            временное проживание             -              -   
2            временное проживание             -              -   
3            временное проживание             -              -   
4                               -             +              +   

  Снят/необходимо снять с контроля  
0                   Нет информации  
1                   Нет информации  
2                   Нет информации  
3                   Нет информации  
4                 Снято с контроля

In [30]:
df['Снят/необходимо снять с контроля'].unique()

array(['Нет информации', 'Снято с контроля',
       'Некапитальные. Направлено на снятие'], dtype=object)

In [31]:
# рассмотрим теперь только те данные у которых нет проблематики (удаляю те у которых 'Снят/необходимо ...' == 'Снято с контроля','Некапитальные. Направлено на снятие'; и 'Поселение\n(без Москвы)' == 'Москва')

data_filtered = df[~(
    (df['Снят/необходимо снять с контроля'].isin(['Снято с контроля', 'Некапитальные. Направлено на снятие'])) | 
    (df['Поселение\n(без Москвы)'] == 'Москва')
)]

data_filtered.shape


(15774, 12)

In [32]:
data_filtered['Кад. номер ОКС'].unique()

array(['-', '+'], dtype=object)

### Сначала построю бинарную регрессию

In [33]:
# создам бинарную переменную оформлен/не оформлен:
data_filtered['Статус оформления'] = ((data_filtered['Кад. номер ЗУ'] == '+') & (data_filtered['Кад. номер ОКС'] == '+')).astype(int)
data_filtered.head()

/var/folders/7d/vfknycw97ws8rbtkkch8f8vm0000gn/T/ipykernel_5908/2426179932.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_filtered['Статус оформления'] = ((data_filtered['Кад. номер ЗУ'] == '+') & (data_filtered['Кад. номер ОКС'] == '+')).astype(int)


№ п/п          Координата яндекс Название посёлка   Поселение  \
0     1  55.614630070,37.232309183        кв-л № 73  Внуковское   
1     2  55.614013325,37.230067327        кв-л № 73  Внуковское   
2     3  55.614692112,37.230466380        кв-л № 73  Внуковское   
3     4  55.615425187,37.229852989        кв-л № 73  Внуковское   
5     6  55.617449028,37.269484212   деревня Ликова  Внуковское   

  Поселение\n(без Москвы) Плодадь поселения (км²)  \
0              Внуковское                    25.6   
1              Внуковское                    25.6   
2              Внуковское                    25.6   
3              Внуковское                    25.6   
5              Внуковское                    25.6   

  Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)  \
0                                               32.4                         
1                                               32.4                         
2                                               32.4                         
3                                               32.4                         
5                                               32.4                         

  Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)  \
0                                                CНТ          
1                                                CНТ          
2                                                CНТ          
3                                                CНТ          
5                                            деревня          

  Постоянное/временное проживание Кад. номер ЗУ Кад. номер ОКС  \
0            временное проживание             -              -   
1            временное проживание             -              -   
2            временное проживание             -              -   
3            временное проживание             -              -   
5           постоянное проживание             -              +   

  Снят/необходимо снять с контроля  Статус оформления  
0                   Нет информации                  0  
1                   Нет информации                  0  
2                   Нет информации                  0  
3                   Нет информации                  0  
5                   Нет информации                  0

In [34]:
# посомтрим сколько их чтобы знать предварительно насколько хороша будет работать бинарная регрессия
oformlen_counts = data_filtered['Статус оформления'].value_counts()
print(oformlen_counts)


Статус оформления
0    14684
1     1090
Name: count, dtype: int64


Видим, что оформленных объектов больше чем неоформленных на 13594. Исходя из чего можно сказать, что наша будущая бинарная регрессия будет недостаточно точна

In [35]:
# изменим тип колонки для бинарной регрессии
data_filtered['Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)'] = pd.to_numeric(data_filtered['Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)'], errors='coerce')
data_filtered.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15774 entries, 0 to 24697
Data columns (total 13 columns):
 #   Column                                                                    Non-Null Count  Dtype  
---  ------                                                                    --------------  -----  
 0   № п/п                                                                     15774 non-null  object 
 1   Координата яндекс                                                         15774 non-null  object 
 2   Название посёлка                                                          15774 non-null  object 
 3   Поселение                                                                 15774 non-null  object 
 4   Поселение
(без Москвы)                                                    15774 non-null  object 
 5   Плодадь поселения (км²)                                                   15774 non-null  object 
 6   Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший 

/var/folders/7d/vfknycw97ws8rbtkkch8f8vm0000gn/T/ipykernel_5908/3471288740.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_filtered['Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)'] = pd.to_numeric(data_filtered['Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)'], errors='coerce')


In [36]:
# делим на обучающую и тестовые выборки:
from sklearn.model_selection import train_test_split

X = data_filtered[['Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)']]
y = data_filtered['Статус оформления']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)


In [37]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model.fit(X_train, y_train)
predictions = model.predict(X_test)

In [38]:
from sklearn.metrics import classification_report

print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           0       0.92      1.00      0.96      4377
           1       0.00      0.00      0.00       356

    accuracy                           0.92      4733
   macro avg       0.46      0.50      0.48      4733
weighted avg       0.86      0.92      0.89      4733



/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  

По данным результатам бинарной регрессии можно сказать, что модель очень хорошо справляется с предсказанием класса 0 (Не оформлен) (с точностью 92% и полнотой 100%), но совершенно не определяет класс 1 (Оформлен) (с точностью и полнотой 0%). Это может указывать на несколько возможных проблем или особенностей данных:

1. **Несбалансированность классов:** данные сильно несбалансированными, что видно по разнице в количестве наблюдений для двух классов (4377 против 356). Модели машинного обучения часто испытывают трудности с обучением на несбалансированных данных, так как они склонны предсказывать более многочисленный класс

2. **Недостаточная дискриминационная способность признаков:** используемые признаки (в данном случае — расстояние от старой Москвы) не обладают достаточной дискриминационной способностью для различения двух классов эффективно

Что касается нашей гипотезы:

1. **Валидация гипотезы:** Гипотеза утверждала, что существует взаимосвязь между расстоянием от старой Москвы и оформленностью объекта недвижимости — чем больше расстояние, тем выше шансы, что объект не оформлен. Однако результаты модели показывают, что модель (бинарная регрессия) не смогла адекватно классифицировать объекты как неоформленные (класс 1), поскольку для этого класса получены нулевые значения precision и recall.

2. **Ограниченность данных:** Высокая точность (accuracy) модели в основном обусловлена способностью корректно классифицировать оформленные объекты (класс 0), а не способностью различать оформленные и неоформленные объекты. Это указывает на то, что модель в основном предсказывает наиболее частый класс (оформленные объекты) и не выявляет достаточно информации для классификации неоформленных объектов.

3. **Вывод по гипотезе:** Исходя из данных, нельзя утверждать, что гипотеза подтверждена, поскольку модель не демонстрирует эффективности в предсказании неоформленности объектов на основе расстояния от Москвы

### Раз наша бинарная регрессия не смогла ничего ТОЧНО сказать о нашей гипотезе, напишем тогда мультиномиальную регрессию:

In [39]:
data_filtered['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'].unique()

array(['CНТ', 'деревня', 'посёлок', 'ТИЗ', 'ДСК', 'ТСН', 'ДНП', 'ЖСиЭК',
       'ЖСПК', 'ЖСПКИЗ', 'село', 'СПК', 'КИЗ', 'ТСЖ', 'город',
       'рабочий посёлок', 'ДНТ', 'ДПК', 'ДСПК', 'дачный посёлок, дачи',
       'ДП', 'СТ', 'хутор', 'ЖСКИЗ', 'ОНТ', 'СДТ', 'СК', 'ОПХ', 'ЖСК',
       'ИЖСК', 'ТОО', 'НСТ', 'ПК', 'поселение Сосенское', 'с/т'],
      dtype=object)

Для построения мультиноминальной регрессии, я сначала создала новую переменную для статуса оформления, которая будет иметь три категории в соответствии с описанными условиями в гипотезе. Затем я использую эту переменную как зависимую в мультиноминальной регрессионной модели. 
Вот как я закодирую данную новую переменную:

+ Оформлен ЗУ и не оформлен ОКС (кодируем как 0)
+ Не оформлен ЗУ, оформлен ОКС (кодируем как 1)
+ Ничего не оформлено (кодируем как 2)

In [40]:
def categorize_status(row):
    if row['Кад. номер ЗУ'] == '+' and row['Кад. номер ОКС'] == '-':
        return 0  # Оформлен ЗУ и не оформлен ОКС
    elif row['Кад. номер ЗУ'] == '-' and row['Кад. номер ОКС'] == '+':
        return 1  # Не оформлен ЗУ, оформлен ОКС
    return 2  # Ничего не оформлено

data_filtered['Мультиноминальный статус'] = data_filtered.apply(categorize_status, axis=1)


/var/folders/7d/vfknycw97ws8rbtkkch8f8vm0000gn/T/ipykernel_5908/2834045227.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_filtered['Мультиноминальный статус'] = data_filtered.apply(categorize_status, axis=1)


In [41]:
# посомтрим сколько их чтобы знать предварительно насколько хороша будет работать бинарная регрессия
oformlen_counts = data_filtered['Мультиноминальный статус'].value_counts()
print(oformlen_counts)


Мультиноминальный статус
0    8989
2    5188
1    1597
Name: count, dtype: int64


In [42]:
# Посмотрим как выглядит наш датасет:
data_filtered.head()

№ п/п          Координата яндекс Название посёлка   Поселение  \
0     1  55.614630070,37.232309183        кв-л № 73  Внуковское   
1     2  55.614013325,37.230067327        кв-л № 73  Внуковское   
2     3  55.614692112,37.230466380        кв-л № 73  Внуковское   
3     4  55.615425187,37.229852989        кв-л № 73  Внуковское   
5     6  55.617449028,37.269484212   деревня Ликова  Внуковское   

  Поселение\n(без Москвы) Плодадь поселения (км²)  \
0              Внуковское                    25.6   
1              Внуковское                    25.6   
2              Внуковское                    25.6   
3              Внуковское                    25.6   
5              Внуковское                    25.6   

   Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)  \
0                                               32.4                          
1                                               32.4                          
2                                               32.4                          
3                                               32.4                          
5                                               32.4                          

  Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)  \
0                                                CНТ          
1                                                CНТ          
2                                                CНТ          
3                                                CНТ          
5                                            деревня          

  Постоянное/временное проживание Кад. номер ЗУ Кад. номер ОКС  \
0            временное проживание             -              -   
1            временное проживание             -              -   
2            временное проживание             -              -   
3            временное проживание             -              -   
5           постоянное проживание             -              +   

  Снят/необходимо снять с контроля  Статус оформления  \
0                   Нет информации                  0   
1                   Нет информации                  0   
2                   Нет информации                  0   
3                   Нет информации                  0   
5                   Нет информации                  0   

   Мультиноминальный статус  
0                         2  
1                         2  
2                         2  
3                         2  
5                         1

In [43]:
# теперь строим мультиноминальную регрессию:
from sklearn.linear_model import LogisticRegression

X = data_filtered[['Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)']]
y = data_filtered['Мультиноминальный статус']

model = LogisticRegression(multi_class='multinomial', solver='lbfgs')
model.fit(X, y)
predictions = model.predict(X)


In [44]:
# отчет о классификации:
from sklearn.metrics import classification_report
print(classification_report(y, predictions))

              precision    recall  f1-score   support

           0       0.57      1.00      0.73      8989
           1       0.00      0.00      0.00      1597
           2       0.00      0.00      0.00      5188

    accuracy                           0.57     15774
   macro avg       0.19      0.33      0.24     15774
weighted avg       0.32      0.57      0.41     15774



/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  

In [45]:
# коэффиценты регрессии:
list(model.coef_)

[array([0.00824106]), array([-0.01568404]), array([0.00744298])]

In [46]:
# матрица ошибок:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y, predictions)
print(cm)


[[8989    0    0]
 [1597    0    0]
 [5188    0    0]]


Исходя из данных результатов мультиноминальной регрессии, я сделала следующие выводы:

1. **Оценка классификации:**
   - Модель имеет хорошую точность (precision) и полноту (recall) только для класса 0, что указывает на то, что она успешно идентифицирует случаи, когда ЗУ оформлен, а ОКС не оформлен.
   - Классы 1 и 2 (не оформлен ЗУ, оформлен ОКС и ничего не оформлено, соответственно) не идентифицируются вообще. Модель не смогла правильно классифицировать ни одного случая для этих классов, что отражено в нулевых значениях precision и recall.

2. **Коэффициенты регрессии:**
   - Коэффициент для класса 0 положительный, что указывает на положительное влияние расстояния на вероятность оформления ЗУ без оформления ОКС.
   - Отрицательный коэффициент для класса 1 предполагает, что увеличение расстояния уменьшает вероятность оформления ОКС без оформления ЗУ, но модель не смогла подтвердить эту зависимость на практике.
   - Коэффициент для класса 2 близок к коэффициенту класса 0, но, поскольку модель не смогла правильно идентифицировать случаи для класса 2, трудно сделать определенные выводы по этому коэффициенту.

3. **Матрица ошибок:**
   - Вся диагональ матрицы ошибок за исключением первой ячейки заполнена нулями, что говорит о том, что модель предсказывает только класс 0 и игнорирует классы 1 и 2.

Выводы:
- Модель эффективно распознает только объекты из класса 0 (оформлен ЗУ, не оформлен ОКС) и не распознает объекты из классов 1 и 2. Это может свидетельствовать о нехватке информативности предиктора (расстояния от Москвы) для различения этих специфических статусов оформления или о необходимости дополнительного анализа и возможно применения других методов или дополнительных предикторов для более эффективной классификации.
- Результаты не позволяют подтвердить или опровергнуть исходную гипотезу полностью, так как модель не смогла адекватно классифицировать все интересующие нас категории статусов оформления.

Исходя из результатов бинарной и мультиноминальной регрессий, можно сделать определённые наблюдения относительно нашей гипотезы. Что я смогла извлечь из результатов:

1. **Бинарная регрессия:** Показала, что модель достаточно хорошо идентифицирует один класс (в нашем случае "оформленные" объекты), но совершенно не распознаёт другой ("неоформленные"). Это говорит о том, что если бы гипотеза утверждала, что существует значимая связь между расстоянием и оформленностью ЗУ и ОКС для оформленных объектов, то можно было бы говорить о частичном подтверждении. Но поскольку модель не смогла идентифицировать неоформленные объекты, полное подтверждение гипотезы невозможно.

2. **Мультиноминальная регрессия:** Подтвердила способность модели определить только объекты класса, где ЗУ оформлено, а ОКС нет. Однако она не смогла адекватно классифицировать объекты двух других категорий. Это означает, что если гипотеза касалась только взаимосвязи для конкретного типа оформления (ЗУ оформлено, ОКС нет), можно говорить о частичном подтверждении. Но в целом, поскольку анализ не охватывает все аспекты гипотезы, полноценное подтверждение или опровержение гипотезы невозможно.

Итак, можно сказать, что данные частично подтверждают гипотезу только в отношении определённой категории оформления (ЗУ оформлено, ОКС нет) и не позволяют сделать выводы о других категориях. Для более полного анализа потребуются дополнительные данные или другие методы анализа, которые могли бы лучше различать разные категории статуса оформления.

### Подытожим

Исходя из результатов бинарной и мультиноминальной регрессий, можно сделать определённые наблюдения относительно гипотезы №1:

1. **Бинарная регрессия:** Показала, что модель достаточно хорошо идентифицирует один класс (в нашем случае, "оформленные" объекты), но совершенно не распознаёт другой ("неоформленные"). Это говорит о том, что если бы гипотеза утверждала, что существует значимая связь между расстоянием и оформленностью ЗУ и ОКС для оформленных объектов, то можно было бы говорить о частичном подтверждении. Но поскольку модель не смогла идентифицировать неоформленные объекты, полное подтверждение гипотезы невозможно.

2. **Мультиноминальная регрессия:** Подтвердила способность модели определить только объекты класса, где ЗУ оформлено, а ОКС нет. Однако она не смогла адекватно классифицировать объекты двух других категорий. Это означает, что если гипотеза касалась только взаимосвязи для конкретного типа оформления (ЗУ оформлено, ОКС нет), можно говорить о частичном подтверждении. Но в целом, поскольку анализ не охватывает все аспекты гипотезы, полноценное подтверждение или опровержение гипотезы невозможно.

Итак, можно сказать, что данные частично подтверждают гипотезу только в отношении определённой категории оформления (ЗУ оформлено, ОКС нет) и не позволяют сделать выводы о других категориях

## Гипотеза №2

### На территории СНТ/ОНТ больше не оформленных объектов недвижимости по сравнению с домами в деревнях
+ Зависимая переменная: оформлен/ не оформлен
+ Предиктор: расположение в СНТ/ОНТ или в деревне

In [47]:
data_filtered['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'].unique()

array(['CНТ', 'деревня', 'посёлок', 'ТИЗ', 'ДСК', 'ТСН', 'ДНП', 'ЖСиЭК',
       'ЖСПК', 'ЖСПКИЗ', 'село', 'СПК', 'КИЗ', 'ТСЖ', 'город',
       'рабочий посёлок', 'ДНТ', 'ДПК', 'ДСПК', 'дачный посёлок, дачи',
       'ДП', 'СТ', 'хутор', 'ЖСКИЗ', 'ОНТ', 'СДТ', 'СК', 'ОПХ', 'ЖСК',
       'ИЖСК', 'ТОО', 'НСТ', 'ПК', 'поселение Сосенское', 'с/т'],
      dtype=object)

In [48]:
data_filtered.head()

№ п/п          Координата яндекс Название посёлка   Поселение  \
0     1  55.614630070,37.232309183        кв-л № 73  Внуковское   
1     2  55.614013325,37.230067327        кв-л № 73  Внуковское   
2     3  55.614692112,37.230466380        кв-л № 73  Внуковское   
3     4  55.615425187,37.229852989        кв-л № 73  Внуковское   
5     6  55.617449028,37.269484212   деревня Ликова  Внуковское   

  Поселение\n(без Москвы) Плодадь поселения (км²)  \
0              Внуковское                    25.6   
1              Внуковское                    25.6   
2              Внуковское                    25.6   
3              Внуковское                    25.6   
5              Внуковское                    25.6   

   Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)  \
0                                               32.4                          
1                                               32.4                          
2                                               32.4                          
3                                               32.4                          
5                                               32.4                          

  Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)  \
0                                                CНТ          
1                                                CНТ          
2                                                CНТ          
3                                                CНТ          
5                                            деревня          

  Постоянное/временное проживание Кад. номер ЗУ Кад. номер ОКС  \
0            временное проживание             -              -   
1            временное проживание             -              -   
2            временное проживание             -              -   
3            временное проживание             -              -   
5           постоянное проживание             -              +   

  Снят/необходимо снять с контроля  Статус оформления  \
0                   Нет информации                  0   
1                   Нет информации                  0   
2                   Нет информации                  0   
3                   Нет информации                  0   
5                   Нет информации                  0   

   Мультиноминальный статус  
0                         2  
1                         2  
2                         2  
3                         2  
5                         1

In [49]:
# для начала добавим характерный столбец:
data_filtered['Статус СНТ/ОНТ'] = data_filtered['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'].apply(lambda x: 1 if x in ['СНТ', 'ОНТ', 'деревня'] else 0)

oformlen_counts = data_filtered['Статус СНТ/ОНТ'].value_counts()
print(oformlen_counts)

Статус СНТ/ОНТ
0    11038
1     4736
Name: count, dtype: int64


/var/folders/7d/vfknycw97ws8rbtkkch8f8vm0000gn/T/ipykernel_5908/3276373423.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_filtered['Статус СНТ/ОНТ'] = data_filtered['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'].apply(lambda x: 1 if x in ['СНТ', 'ОНТ', 'деревня'] else 0)


In [50]:
data_filtered.head()

№ п/п          Координата яндекс Название посёлка   Поселение  \
0     1  55.614630070,37.232309183        кв-л № 73  Внуковское   
1     2  55.614013325,37.230067327        кв-л № 73  Внуковское   
2     3  55.614692112,37.230466380        кв-л № 73  Внуковское   
3     4  55.615425187,37.229852989        кв-л № 73  Внуковское   
5     6  55.617449028,37.269484212   деревня Ликова  Внуковское   

  Поселение\n(без Москвы) Плодадь поселения (км²)  \
0              Внуковское                    25.6   
1              Внуковское                    25.6   
2              Внуковское                    25.6   
3              Внуковское                    25.6   
5              Внуковское                    25.6   

   Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)  \
0                                               32.4                          
1                                               32.4                          
2                                               32.4                          
3                                               32.4                          
5                                               32.4                          

  Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)  \
0                                                CНТ          
1                                                CНТ          
2                                                CНТ          
3                                                CНТ          
5                                            деревня          

  Постоянное/временное проживание Кад. номер ЗУ Кад. номер ОКС  \
0            временное проживание             -              -   
1            временное проживание             -              -   
2            временное проживание             -              -   
3            временное проживание             -              -   
5           постоянное проживание             -              +   

  Снят/необходимо снять с контроля  Статус оформления  \
0                   Нет информации                  0   
1                   Нет информации                  0   
2                   Нет информации                  0   
3                   Нет информации                  0   
5                   Нет информации                  0   

   Мультиноминальный статус  Статус СНТ/ОНТ  
0                         2               0  
1                         2               0  
2                         2               0  
3                         2               0  
5                         1               1

In [51]:
# обучаем модель:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

X = data_filtered[['Статус СНТ/ОНТ']]
y = data_filtered['Статус оформления']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

binary_model = LogisticRegression()
binary_model.fit(X_train, y_train)
binary_predictions = binary_model.predict(X_test)


In [52]:
print(classification_report(y_test, binary_predictions))


              precision    recall  f1-score   support

           0       0.92      1.00      0.96      4377
           1       0.00      0.00      0.00       356

    accuracy                           0.92      4733
   macro avg       0.46      0.50      0.48      4733
weighted avg       0.86      0.92      0.89      4733



/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  

Результаты показывают, что модель имеет высокую общую точность (accuracy), однако она снова демонстрирует низкую эффективность в идентификации объектов класса 1 ("неоформленные" объекты):

1. **Высокая точность для класса 0 (оформленные объекты):**
   - Модель эффективно предсказывает класс 0 с точностью (precision) 0.92 и полнотой (recall) 1.00. Это означает, что модель очень хорошо идентифицирует оформленные объекты.

2. **Низкая производительность для класса 1 (неоформленные объекты):**
   - Модель не смогла корректно идентифицировать неоформленные объекты (класс 1), что видно из нулевых значений точности и полноты. Это означает, что модель не предсказала ни одного неоформленного объекта правильно.

3. **Общая точность модели:**
   - Общая точность (accuracy) модели составляет 0.92, но это число в основном отражает способность модели идентифицировать только оформленные объекты из-за дисбаланса классов в данных.

Исходя из этих результатов, сложно подтвердить вторую гипотезу, которая предполагает, что на территории СНТ/ОНТ больше не оформленных объектов недвижимости по сравнению с домами в деревнях.

### теперь построим мультиноминальную регрессию:

In [53]:
X_multinomial = data_filtered[['Статус СНТ/ОНТ']]  # Независимая переменная
y_multinomial = data_filtered['Мультиноминальный статус']  # Зависимая переменная

# Разделение данных на обучающую и тестовую выборки
X_train_multinomial, X_test_multinomial, y_train_multinomial, y_test_multinomial = train_test_split(
    X_multinomial, y_multinomial, test_size=0.3, random_state=42)

# Построение и обучение модели мультиноминальной логистической регрессии
multinomial_model = LogisticRegression(multi_class='multinomial', solver='lbfgs')
multinomial_model.fit(X_train_multinomial, y_train_multinomial)
multinomial_predictions = multinomial_model.predict(X_test_multinomial)

multinomial_classification_report = classification_report(y_test_multinomial, multinomial_predictions)
print(multinomial_classification_report)


              precision    recall  f1-score   support

           0       0.57      1.00      0.73      2700
           1       0.00      0.00      0.00       458
           2       0.00      0.00      0.00      1575

    accuracy                           0.57      4733
   macro avg       0.19      0.33      0.24      4733
weighted avg       0.33      0.57      0.41      4733



/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  

Результаты мультиноминальной регрессии, показывают следующее:

1. **Класс 0 (предположительно, "оформлен ЗУ и не оформлен ОКС"):**
   - Модель хорошо справляется с идентификацией этого класса, достигая точности (precision) 0.57 и полноты (recall) 1.00. Это означает, что модель идентифицирует большинство или все случаи этого класса, но при этом допускает некоторое количество ложноположительных результатов.

2. **Класс 1 (предположительно, "не оформлен ЗУ, оформлен ОКС") и Класс 2 (предположительно, "ничего не оформлено"):**
   - Модель не смогла идентифицировать случаи, принадлежащие к классам 1 и 2, что показывают нулевые значения precision и recall. Это указывает на неспособность модели различать эти категории статуса оформления на основе имеющихся данных.

3. **Общая точность (accuracy):**
   - Общая точность модели составляет 0.57, что в основном обусловлено способностью модели правильно идентифицировать случаи класса 0.

Исходя из этих результатов, можно сделать вывод, что модель эффективно определяет статус оформления только для одной категории (оформлен ЗУ и не оформлен ОКС). Для других категорий статуса оформления модель не показывает эффективности

Исходя из результатов как бинарной, так и мультиноминальной регрессии, я сделала следующие выводы относительно второй гипотезы:

**Гипотеза 2:** На территории СНТ/ОНТ больше не оформленных объектов недвижимости по сравнению с домами в деревнях.

1. **Бинарная регрессия:**
   - Модель показала высокую точность для одного класса, но эта точность была связана с правильным определением оформленных объектов, в то время как неоформленные объекты модель определить не смогла.
   - Не было выявлено достаточных доказательств для подтверждения гипотезы, поскольку модель не продемонстрировала способность различать оформленные и неоформленные объекты на основе их расположения (в СНТ/ОНТ или деревне).

2. **Мультиноминальная регрессия:**
   - Мультиноминальная модель успешно идентифицировала только одну категорию статуса оформления, не продемонстрировав способность различать все три предполагаемые категории.
   - Так как не было установлено чёткого различия между статусами оформления в зависимости от расположения объекта (СНТ/ОНТ против деревни), результаты не подтверждают гипотезу о том, что неоформленность зависит от типа населённого пункта.

**Общие выводы:**

На основе текущих результатов нельзя однозначно подтвердить гипотезу о том, что на территории СНТ/ОНТ больше не оформленных объектов недвижимости, чем в деревнях. Модели не показали значимой зависимости между типом населённого пункта и статусом оформления объектов недвижимости

In [54]:
data_filtered.to_excel('Измененнаябд.xlsx', index=False)


In [55]:
data_filtered.head()

№ п/п          Координата яндекс Название посёлка   Поселение  \
0     1  55.614630070,37.232309183        кв-л № 73  Внуковское   
1     2  55.614013325,37.230067327        кв-л № 73  Внуковское   
2     3  55.614692112,37.230466380        кв-л № 73  Внуковское   
3     4  55.615425187,37.229852989        кв-л № 73  Внуковское   
5     6  55.617449028,37.269484212   деревня Ликова  Внуковское   

  Поселение\n(без Москвы) Плодадь поселения (км²)  \
0              Внуковское                    25.6   
1              Внуковское                    25.6   
2              Внуковское                    25.6   
3              Внуковское                    25.6   
5              Внуковское                    25.6   

   Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)  \
0                                               32.4                          
1                                               32.4                          
2                                               32.4                          
3                                               32.4                          
5                                               32.4                          

  Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)  \
0                                                CНТ          
1                                                CНТ          
2                                                CНТ          
3                                                CНТ          
5                                            деревня          

  Постоянное/временное проживание Кад. номер ЗУ Кад. номер ОКС  \
0            временное проживание             -              -   
1            временное проживание             -              -   
2            временное проживание             -              -   
3            временное проживание             -              -   
5           постоянное проживание             -              +   

  Снят/необходимо снять с контроля  Статус оформления  \
0                   Нет информации                  0   
1                   Нет информации                  0   
2                   Нет информации                  0   
3                   Нет информации                  0   
5                   Нет информации                  0   

   Мультиноминальный статус  Статус СНТ/ОНТ  
0                         2               0  
1                         2               0  
2                         2               0  
3                         2               0  
5                         1               1

In [56]:
list(data_filtered.columns)

['№ п/п',
 'Координата яндекс',
 'Название посёлка',
 'Поселение',
 'Поселение\n(без Москвы)',
 'Плодадь поселения (км²)',
 'Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)',
 'Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)',
 'Постоянное/временное проживание',
 'Кад. номер ЗУ',
 'Кад. номер ОКС',
 'Снят/необходимо снять с контроля',
 'Статус оформления',
 'Мультиноминальный статус',
 'Статус СНТ/ОНТ']

In [57]:
data_filtered.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15774 entries, 0 to 24697
Data columns (total 15 columns):
 #   Column                                                                    Non-Null Count  Dtype  
---  ------                                                                    --------------  -----  
 0   № п/п                                                                     15774 non-null  object 
 1   Координата яндекс                                                         15774 non-null  object 
 2   Название посёлка                                                          15774 non-null  object 
 3   Поселение                                                                 15774 non-null  object 
 4   Поселение
(без Москвы)                                                    15774 non-null  object 
 5   Плодадь поселения (км²)                                                   15774 non-null  object 
 6   Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший 

In [58]:
print(data_filtered.isnull().sum())

№ п/п                                                                       0
Координата яндекс                                                           0
Название посёлка                                                            0
Поселение                                                                   0
Поселение\n(без Москвы)                                                     0
Плодадь поселения (км²)                                                     0
Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)    0
Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)                   0
Постоянное/временное проживание                                             0
Кад. номер ЗУ                                                               0
Кад. номер ОКС                                                              0
Снят/необходимо снять с контроля                                            0
Статус оформления                                               

Предикторы: В качестве предикторов я выбрала 'Плодадь поселения (км²)' и 'Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)'.

Зависимая переменная: В качестве зависимой переменной я выбрала 'Статус СНТ/ОНТ', так как это является целевой переменной, которую мы хотим предсказывать

In [59]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score


data_filtered.drop(['№ п/п', 'Координата яндекс', 'Название посёлка', 'Поселение\n(без Москвы)', 'Постоянное/временное проживание', 'Кад. номер ЗУ', 'Кад. номер ОКС', 'Снят/необходимо снять с контроля', 'Мультиноминальный статус'], axis=1, inplace=True)
data_filtered['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'] = data_filtered['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'].map({'СНТ': 1, 'ОНТ': 1, 'ДНТ': 1, 'город': 0, 'деревня': 0})

X = data_filtered[['Плодадь поселения (км²)', 'Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)', 'Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)']]
y = data_filtered['Статус СНТ/ОНТ']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

predictions = model.predict(X_test)
accuracy = accuracy_score(y_test, predictions)
print("Accuracy:", accuracy)


Accuracy: 0.9996830427892235


/var/folders/7d/vfknycw97ws8rbtkkch8f8vm0000gn/T/ipykernel_5908/1794932698.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_filtered.drop(['№ п/п', 'Координата яндекс', 'Название посёлка', 'Поселение\n(без Москвы)', 'Постоянное/временное проживание', 'Кад. номер ЗУ', 'Кад. номер ОКС', 'Снят/необходимо снять с контроля', 'Мультиноминальный статус'], axis=1, inplace=True)
/var/folders/7d/vfknycw97ws8rbtkkch8f8vm0000gn/T/ipykernel_5908/1794932698.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_filtered['Статус населенного пункта (СНТ/ОНТ/ДНТ или гор

In [60]:
feature_importances = model.feature_importances_
print("Важность признаков:")
print("Плодадь поселения (км²):", feature_importances[0])
print("Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь):", feature_importances[1])


Важность признаков:
Плодадь поселения (км²): 0.05384213634997783
Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь): 0.06286741396346487


Исходя из результатов модели случайного леса, я сделала следующие выводы относительно двух гипотез:

1. **Есть взаимосвязь между расстоянием от Москвы и оформлением объекта недвижимости**: Коэффициент для 'Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)' равен 0.0629, что означает, что этот признак оказывает наибольшее влияние на модель из всех рассматриваемых признаков. Увеличение удаленности от Москвы приводит к увеличению вероятности объекта находиться в СНТ/ОНТ. Это подтверждает первую гипотезу.

2. **На территории СНТ/ОНТ больше неоформленных объектов недвижимости по сравнению с домами в деревнях**: Коэффициент для 'Плодадь поселения (км²)' равен 0.0538, что также означает, что этот признак оказывает влияние на модель, хоть и не такое сильное, как удаленность от Москвы. Однако, важность этого признака указывает на то, что площадь поселения может быть также важным фактором. Таким образом эта информация подтверждает вторую гипотезу, хотя другие факторы также могут играть роль в оформлении недвижимости на территории СНТ/ОНТ.

Общая точность модели составляет 99.97%, что является очень высоким значением

In [61]:
data_filtered.head()

Поселение Плодадь поселения (км²)  \
0  Внуковское                    25.6   
1  Внуковское                    25.6   
2  Внуковское                    25.6   
3  Внуковское                    25.6   
5  Внуковское                    25.6   

   Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)  \
0                                               32.4                          
1                                               32.4                          
2                                               32.4                          
3                                               32.4                          
5                                               32.4                          

   Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)  \
0                                                NaN           
1                                                NaN           
2                                                NaN           
3                                                NaN           
5                                                0.0           

   Статус оформления  Статус СНТ/ОНТ  
0                  0               0  
1                  0               0  
2                  0               0  
3                  0               0  
5                  0               1

Теперь построим модель которая будет предсказывать оформлен обьект или нет:

In [63]:
dataframe = pd.read_excel('Измененнаябд.xlsx')
dataframe.head()

№ п/п          Координата яндекс Название посёлка   Поселение  \
0     1  55.614630070,37.232309183        кв-л № 73  Внуковское   
1     2  55.614013325,37.230067327        кв-л № 73  Внуковское   
2     3  55.614692112,37.230466380        кв-л № 73  Внуковское   
3     4  55.615425187,37.229852989        кв-л № 73  Внуковское   
4     6  55.617449028,37.269484212   деревня Ликова  Внуковское   

  Поселение\n(без Москвы)  Плодадь поселения (км²)  \
0              Внуковское                     25.6   
1              Внуковское                     25.6   
2              Внуковское                     25.6   
3              Внуковское                     25.6   
4              Внуковское                     25.6   

   Удаленность от Москвы (км). Точка отсчета центр Москвы (крадчайший путь)  \
0                                               32.4                          
1                                               32.4                          
2                                               32.4                          
3                                               32.4                          
4                                               32.4                          

  Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)  \
0                                                CНТ          
1                                                CНТ          
2                                                CНТ          
3                                                CНТ          
4                                            деревня          

  Постоянное/временное проживание Кад. номер ЗУ Кад. номер ОКС  \
0            временное проживание             -              -   
1            временное проживание             -              -   
2            временное проживание             -              -   
3            временное проживание             -              -   
4           постоянное проживание             -              +   

  Снят/необходимо снять с контроля  Статус оформления  \
0                   Нет информации                  0   
1                   Нет информации                  0   
2                   Нет информации                  0   
3                   Нет информации                  0   
4                   Нет информации                  0   

   Мультиноминальный статус  Статус СНТ/ОНТ  
0                         2               0  
1                         2               0  
2                         2               0  
3                         2               0  
4                         1               1

In [65]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report

dataframe = dataframe.drop(['№ п/п', 'Координата яндекс', 'Название посёлка', 'Поселение', 'Поселение\n(без Москвы)', 'Кад. номер ЗУ', 'Кад. номер ОКС', 'Снят/необходимо снять с контроля', 'Мультиноминальный статус', 'Статус СНТ/ОНТ'], axis=1)

# Преобразование категориальных переменных в числовые
dataframe['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)'] = pd.Categorical(dataframe['Статус населенного пункта (СНТ/ОНТ/ДНТ или город/деревня)']).codes
dataframe['Постоянное/временное проживание'] = pd.Categorical(dataframe['Постоянное/временное проживание']).codes


In [66]:
X = dataframe.drop('Статус оформления', axis=1)
y = dataframe['Статус оформления']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [67]:
model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [68]:
predictions = model.predict(X_test)
accuracy = accuracy_score(y_test, predictions)
print("Точность:", accuracy)
print(classification_report(y_test, predictions))


Точность: 0.9236133122028526
              precision    recall  f1-score   support

           0       0.92      1.00      0.96      2913
           1       0.60      0.01      0.02       242

    accuracy                           0.92      3155
   macro avg       0.76      0.51      0.49      3155
weighted avg       0.90      0.92      0.89      3155



Результаты показывают, что модель достигла общей точности (accuracy) в 92.36%, что кажется весьма обнадёживающим на первый взгляд. Однако, детальный анализ метрик precision, recall и f1-score для каждого класса выявляет некоторые важные нюансы:

- Для класса **0 (неоформленные объекты)** модель показывает высокую точность (precision) в 92% и полноту (recall) в 100%, что приводит к f1-score в 96%. Это говорит о том, что модель очень хорошо справляется с определением неоформленных объектов, правильно идентифицируя почти все реальные случаи из тестового набора данных.

- В отличие от этого, для класса **1 (оформленные объекты)** модель демонстрирует значительно худшие результаты: точность составляет всего 60%, а полнота — всего 1%, что ведёт к очень низкому f1-score в 2%. Это означает, что, хотя когда модель предсказывает, что объект оформлен, она делает это сравнительно точно, она практически не способна корректно идентифицировать большинство оформленных объектов, упуская их и ошибочно классифицируя как неоформленные.

Это расхождение между классами, особенно низкая полнота для оформленных объектов, указывает на проблему с балансом классов в данных